boa/src/syntax/lexer/regex.rs

```diff
@@ -0,0 +1,109 @@
+///
+/// [spec]: https://www.ecma-international.org/ecma-262/#sec-literals-regular-expression-literals
+/// [mdn]:
+#[derive(Debug, Clone, Copy)]
+pub(super) struct RegexLiteral;
+
+impl RegexLiteral {
+    /// Creates a new regex literal lexer.
+    pub(super) fn new() -> Self {
+        Self {}
+    }
+}
+
+impl<R> Tokenizer<R> for RegexLiteral {
+    fn lex(&mut self, cursor: &mut Cursor<R>, _start_pos: Position) -> Result<Token, Error>
+    where
+        R: Read,
+    {
+        let mut body = String::new();
+
+        // Lex RegularExpressionBody.
+        loop {
+            match cursor.next() {
+                None => {
+                    // Abrupt end.
+                    return Err(Error::syntax("Abrupt end, regex not terminated"));
+                }
+                Some(Err(e)) => {
+                    return Err(Error::from(e));
+                }
+                Some(Ok(c)) => {
+                    match c {
+                        '/' => break, // RegularExpressionBody finished.
+                        '\n' | '\r' | '\u{2028}' | '\u{2029}' => {
+                            // Not allowed in Regex literal.
+                            return Err(Error::syntax("Encountered new line during regex"));
+                        }
+                        '\\' => {
+                            // Escape sequence
+                            body.push('\\');
+                            match cursor.next() {
+                                None => {
+                                    // Abrupt end of regex.
+                                    return Err(Error::syntax("Abrupt end, regex not terminated"));
+                                }
+                                Some(Err(_)) => {
+                                    return Err(Error::from(io::Error::new(
+                                        ErrorKind::Interrupted,
+                                        "Failed to peek next character",
+                                    )))
+                                }
+                                Some(Ok(sc)) => {
+                                    match sc {
+                                        '\n' | '\r' | '\u{2028}' | '\u{2029}' => {
+                                            // Not allowed in Regex literal.
+                                            return Err(Error::syntax(
+                                                "Encountered new line during regex",
+                                            ));
+                                        }
+                                        ch => body.push(ch),
+                                    }
+                                }
+                            }
+                        }
+                        _ => body.push(c),
+                    }
+                }
+            }
+        }
+
+        unimplemented!(
+            "Regex handling, requires ability to peek more than a single character ahead"
+        );
```

Comment:
So, the thing here is that this depends on the goal symbol. A `/` in a `Div` goal symbol is always a dividion, no need to check further. A `/` in an `InputElementRegExp` goal symbol is always the start / end of a regular expression, so no need to peek further.

This is the main objective of this change, to allow using goal symbols.

Reply:
Yeah, this code is basically just straight moved over and I haven't really looked at regex much yet so this is still todo